packages/mcp-cloudflare/src/server/lib/mcp-agent.ts

```diff
@@ -0,0 +1,160 @@
+ * that persist across DO hibernation cycles. The DO uses these credentials to authenticate
+ * all Sentry API requests on behalf of the MCP client.
+ */
+
+/**
+ * Sentry MCP Agent - A Durable Object that provides Model Context Protocol access to Sentry.
+ *
+ * Following the pattern from the gist: this class extends McpAgent and mutates props
+ * directly on first request to include URL-extracted constraints. Props are then immutable
+ * throughout the DO's lifetime.
+ */
+class SentryMCPBase extends McpAgent<
+  Env,
+  {
+    constraints?: Constraints;
+  },
+  WorkerProps & {
+    organizationSlug?: string;
+    projectSlug?: string;
+  }
+> {
+  // Create server once in constructor, as per Cloudflare MCP Agent API docs
+  server = new McpServer({
+    name: "Sentry MCP",
+    version: LIB_VERSION,
+  });
+
+  // biome-ignore lint/complexity/noUselessConstructor: Need the constructor to match the durable object types.
+  constructor(state: DurableObjectState, env: Env) {
+    super(state, env);
+  }
+
+  /**
+   * Initialize Durable Object state and MCP server configuration
+   *
+   * Called when the DO is first created or wakes from hibernation.
+   * Configures the MCP server with user authentication and constraint scoping.
+   *
+   * Since props (including constraints) are set on first request and then immutable,
+   * the server configuration remains constant throughout the DO's lifetime.
+   */
+  async init() {
+    if (!this.state?.constraints) {
+      this.setState({
+        constraints: this.props.constraints,
+      });
+    }
+
+    await configureServer({
+      server: this.server,
+      context: {
+        userId: this.props.id,
+        mcpUrl: process.env.MCP_URL,
+        accessToken: this.props.accessToken,
+        constraints: this.state.constraints || {},
+      },
+      onToolComplete: () => {
+        this.ctx.waitUntil(Sentry.flush(2000));
+      },
+    });
+  }
+}
+
+export const SentryMCP = Sentry.instrumentDurableObjectWithSentry(
+  getSentryConfig.partial({
+    initialScope: {
+      tags: {
+        durable_object: true,
+        "mcp.server_version": LIB_VERSION,
+      },
+    },
+  }),
+  SentryMCPBase,
+);
+
+export default {
+  async fetch(request: Request, env: Env, ctx: ExecutionContext) {
+    const url = new URL(request.url);
+
+    if (url.pathname === "/sse" || url.pathname === "/sse/message") {
+      return SentryMCP.serveSSE("/sse").fetch(request, env, ctx);
+    }
+
+    const pattern = new URLPattern({ pathname: "/mcp/:org?/:project?" });
+    const result = pattern.exec(url);
+    if (result) {
+      const { groups } = result.pathname;
+
+      const organizationSlug = groups?.org ?? "";
+      const projectSlug = groups?.project ?? "";
+
+      // Verify access to org/project using OAuth token
+      const verification = await verifyConstraintsAccess(
+        { organizationSlug, projectSlug },
+        {
+          // @ts-ignore props is provided by OAuth provider → agents library
```

Comment:
not totally sure how you're suppsoed to type props, as ExecutionContext has it typed as any